File: sdks/rs/src/config/security.rs
#[derive(Clone, Debug)]
pub struct SecurityConfig {
    /// Always redact these keys regardless of redactor config
    pub force_redact_keys: Vec<String>,
    /// Enable PII detection patterns
    pub pii_detection: bool,
    /// Custom PII patterns (regex)
    pub pii_patterns: Vec<String>,
    /// Hash algorithm for sensitive values ("sha256", "sha512")
    pub hash_algorithm: String,
    /// Maximum bytes for a single field value
    pub max_field_bytes: usize,
    /// Maximum number of attributes per event
    pub max_attr_count: usize,
    /// Drop events that exceed size limits instead of truncating
    pub drop_oversized_events: bool,
}

impl Default for SecurityConfig {
    fn default() -> Self {
        Self {
            force_redact_keys: Vec::new(),
            pii_detection: false,
            pii_patterns: Vec::new(),
            hash_algorithm: "sha256".to_string(),
            max_field_bytes: 0,
            max_attr_count: 0,
            drop_oversized_events: false,
        }
    }
}

File: sdks/js/src/config/security.ts
export interface SecurityConfig {
  redactByDefault?: boolean;
  allowPII?: boolean;
  maxFieldBytes?: number;
  maxEventBytes?: number;
  maxAttrCount?: number;
  dropOversizedEvents?: boolean;
}

const DEFAULT_SECURITY: Required<SecurityConfig> = {
  redactByDefault: false,
  allowPII: false,
  maxFieldBytes: 4096,
  maxEventBytes: 262144,
  maxAttrCount: 512,
  dropOversizedEvents: true,
};

export class SecurityLimiter {
  private config: Required<SecurityConfig>;

  constructor(config?: SecurityConfig) {
    this.config = { ...DEFAULT_SECURITY, ...config };
  }

  /** Returns true if the event should be dropped. */
  shouldDrop(encoded: string, attrCount: number): boolean {
    if (!this.config.dropOversizedEvents) return false;
    if (encoded.length > this.config.maxEventBytes) return true;
    if (attrCount > this.config.maxAttrCount) return true;
    return false;
  }

  /** Check if a single field value exceeds max field bytes. */
  isFieldOversized(value: string): boolean {
    return value.length > this.config.maxFieldBytes;
  }

  getConfig(): Required<SecurityConfig> {
    return { ...this.config };
  }
}
